### PR TITLE
Build scripts for Packaging and Publishing GitVersion

### PR DESCRIPTION
**Pull request by arturcic**
This is a PR for the issue https://github.com/GitTools/GitVersion/issues/1445.

Some details about the scripts - inspired by the work done in the cake-build project.
There is a powershell core script - bootstrap that runs on all platforms - run.ps1
Added support for building on Azure Pipelines for all 3 platforms
The run.cake script is the one covering the build/testing/packaging and publishing.

The unit tests run on each of the platforms both `dotnet test` as well as using NUnit.

The Build/ Test and Packaging run on each build on each platform so that we could actually publish from any build server.

The Publish is currently enabled from AppVeyor. 

There are 2 type of releases - from GitTools/GitVersion repository master branch. If it is tagged - the release is a stable release. If it's not tagged it's a pre-release. 

The pre-release artifacts are published only to AppVeyor/Azure Pipelines. For the Stable release it's also published from AppVeyor to DockerHub, Nuget.org, Chocolatey.org, rubygems.org, and from Travis the Docker images. I decided to enable the Docker build and publish only if it's a stable or pre-release as it takes a lot of time to build them on windows - ~ 30 mins

I have tested pushing for all packages (using self hosted proget server) except for TFS.

There are 4 Docker images - 
2 Windows images - one based on nano with dotnet core runtime image, one based on windowsservercore with full framework 4.7.2
2 Linux images - one based on dotnet core runtime image, one based on mono.

There are several required environment variables that we need to update with corresponding values:

GITHUB_TOKEN
DOCKER_USERNAME
DOCKER_PASSWORD
NUGET_API_KEY
NUGET_API_URL
CHOCOLATEY_API_KEY
CHOCOLATEY_API_URL
TFX_TOKEN
RUBY_GEM_API_KEY

There are also some optional env variables that act as flags to disable some parts of the build script: 
DISABLE_UNIT_TESTS
DISABLE_PUBLISH_GEM
DISABLE_PUBLISH_TFS
DISABLE_PUBLISH_NUGET
DISABLE_PUBLISH_CHOCOLATEY
DISABLE_PUBLISH_DOCKER

Currently the README.md file has a badge for Azure pipelines build status that needs to be updated with the GitTools team's Azure pipelines url when it's set up.

I copied the nuspec files to the root of the repo as well as I have moved the code for ruby gems to a separate folder.

I kept the old build scripts so that you can compare them side by side with the new one.

I wasn't able to test the push of TFS package so I have included the code but it's not reachable at the moment. I'll need some assistance on that.

Please have a look and review it. Let me know if it covers all the scenarios and all the comments are welcomed. 

@JakeGinnivan @dazinator @asbjornu @gep13

### CODE REVIEW

File: run.ps1 [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    if ($path -ne $null) {
+        $newItems = $path.Split(';', [StringSplitOptions]::RemoveEmptyEntries) | Where-Object { "$($_)" -inotlike $VariableToRemove }
+        [Environment]::SetEnvironmentVariable("PATH", [System.String]::Join(';', $newItems), "Process")
+    }
+}
+
+# Get .NET Core CLI path if installed.
+$FoundDotNetCliVersion = $null;
+if (Get-Command dotnet -ErrorAction SilentlyContinue) {
+    $FoundDotNetCliVersion = dotnet --version;
```
[the remark this answers is not in the document]
Comment (arturcic) [reply run.ps1]: that is correct. I chose to install the global.json file so that we can build the current code as is without re-targeting to higher version of .net.

**Comment by arturcic** [comment]
I'll have to check why the Gem packaging is failing when it's a pull request - it's somehow related to the version generated when it's a pull request

File: .travis.yml [hunk 2]
```diff
@@ -1,17 +1,18 @@
...
 os:
   - linux
   - osx
 before_install:
-  - git fetch --unshallow # Travis always does a shallow clone, but GitVersion needs the full history including branches and tags 
+  - git fetch --unshallow # Travis always does a shallow clone, but GitVersion needs the full history including branches and tags
+  - bash <(wget -O - https://raw.githubusercontent.com/PowerShell/PowerShell/master/tools/install-powershell.sh)
 script:
-  - ./build.sh -v Diagnostic
+  - pwsh ./run.ps1 -script build.cake -target Travis
```
[the remark this answers is not in the document]
Comment (arturcic) [reply .travis.yml]: btw the powershell script is using (the latest version) is using Cake global tool. Initially I was thinking to install the tool and call it directly from travis/appveyor/azure-pipelines yaml files, but than decided to keep the bootsrap script to be able download nuget and dotnet for global.json file in the repository

File: run.ps1 [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+$env:CAKE_SETTINGS_SKIPVERIFICATION='true'
+$env:DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
+$env:DOTNET_CLI_TELEMETRY_OPTOUT=1
+
+###########################################################################
+# INSTALL NUGET
+###########################################################################
+
+# Make sure nuget.exe exists.
+$NugetPath = Join-Path $ToolPath "nuget.exe"
```
[the remark this answers is not in the document]
Comment (dazinator) [reply run.ps1]: Got ya, so the unix build agents have `mono` on. Ok cool.

**Comment by asbjornu** [comment]
This looks like a huge improvement over the current build. Would it be possible to have prereleases auto-published to NuGet? Otherwise I think this looks great. Thanks for the work you put into this, @arturcic! Is there anything I can do to help?

**Comment by asbjornu** [comment]
@arturcic: Excellent. I'd like every package repository that supports unstable/pre-releases to be included when we do pre-releases. I don't know much about other package repositories than NuGet and Homebrew, so I can't say authoritatively which repositories should and shouldn't be included other than NuGet, since it supports it. Homebrew does not, so it should be excluded.

I have access to AppVeyor, but I don't have access to Docker, NuGet, Chocolatey, TFX (what's that?) or RubyGems, so I can't generate new credentials. I'm also not entirely sure how to generate a GitHub auth token properly. What access does AppVeyor need and who should own the token?

I've created an Azure DevOps account for GitTools and started setting up a pipeline, but stopped at the template step as I'm unsure whether the suggested ".NET Desktop" template would be a good fit. Perhaps we could hash this out on [Gitter](https://gitter.im/GitTools/GitVersion) to speed up the interaction?

**Comment by JakeGinnivan** [comment]
If you DM me on twitter I can probably grant access to anything. This is awesome, thanks @arturcic 

We should push to docker as a tagged image, and only update latest on stable

**Comment by arturcic** [comment]
> I have access to AppVeyor, but I don't have access to Docker, NuGet, Chocolatey, TFX (what's that?) or RubyGems, so I can't generate new credentials. I'm also not entirely sure how to generate a GitHub auth token properly. What access does AppVeyor need and who should own the token?

tfx-cli is the tool used for packaging/publishing to Visual Studio marketplace, https://github.com/Microsoft/tfs-cli, so basically it's a task that can be used in TFS or Azure DevOps

> @arturcic: Excellent. I'd like every package repository that supports unstable/pre-releases to be included when we do pre-releases.

As far as I know, Nuget, Chocolatey, RubyGems support pre-release, not sure about VS marketplace.
Docker has tags that we can use for pre-releases as well. Regarding Homebrew, we don't support that yet, I was planning to add that after this is merged

**Comment by kaylangan** [comment]
I'm a Program Manager on Azure Pipelines. Let me know if you have any questions or suggestions. I'm happy to help!

**Comment by arturcic** [comment]
> I'm a Program Manager on Azure Pipelines. Let me know if you have any questions or suggestions. I'm happy to help!

Good to know, thanks

**Comment by gep13** [comment]
@arturcic this is an AWESOME pull request.  I have only had a very quick look through the changes, but everything seems legit to me.

@JakeGinnivan were you able to setup access to everything that is required?  If not, let me know what is still outstanding, and I can take a look.

File: build/utils.cake [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    var toolPath = Context.Tools.Resolve("GitReleaseNotes.exe");
+
+    var arguments = new ProcessArgumentBuilder()
+                .Append(workDir.ToString())
+                .Append("/OutputFile")
+                .Append(outputPath.ToString())
+                .Append("/RepoToken")
+                .Append(repoToken);
+
+    StartProcess(toolPath, new ProcessSettings { Arguments = arguments, RedirectStandardOutput = true }, out var redirectedOutput);
```
[the remark this answers is not in the document]
Comment (arturcic) [reply build/utils.cake]: First I have tried to use the cake alias for that, for some reason locally it was working, on build servers - AppVeyor and Azure windows agents, it was crashing, but when calling with StartProcess was working. I guess after we finish with this PR I'll get back to this issue and try to use the alias

File: nuspec/chocolateyInstall.ps1 [hunk 5]
```diff
@@ -0,0 +1 @@
+Generate-BinFile "gfv" "$packageFolder\Tools\GitVersion.exe"
```
[the remark this answers is not in the document]
Comment (arturcic) [reply nuspec/chocolateyInstall.ps1]: I'll change that, thanks

File: build/utils.cake (hunk 4 above)
[the remark this answers is not in the document]
Comment (gep13) [reply build/utils.cake]: Hmm, that is very interesting. Would be great to get to the bottom of that, as it might be that we need to change something in Cake itself (which is where the alias lives).  As you say, this shouldn't stop the PR from moving forward, but something to circle back around to looking at.

[the remark this answers is not in the document]
Comment (arturcic) [reply build/utils.cake]: https://ci.appveyor.com/project/arturcic/gitversion-ee285/build/4.0.0-build-scripts.1+1626.build.267#L205 this is the error when using the alias

[the remark this answers is not in the document]
Comment (gep13) [reply build/utils.cake]: @arturcic would be good to run that with Diagnostic Verbosity to "see" what command that Cake is sending to the command line.  Could be that there is a bug in there.

[the remark this answers is not in the document]
Comment (arturcic) [reply build/utils.cake]: thanks for suggestion, I'll do that as part of a new PR for using the alias

**Comment by gep13** [comment]
@arturcic thanks for making those changes!

**Comment by gep13** [comment]
So, all in all, this looks good to me, an EPIC PR from @arturcic.

@asbjornu @dazinator @JakeGinnivan you guys ok with this getting pulled in?

**Comment by arturcic** [comment]
Quick update on this
- fixed the versioning schema for artifacts, it remains to fix the tfs versioning as it does not acccept pre-releases
- tested the publish to Docker (you can check the dockerhub of gittools). After the merge of this PR I'll update the descriptions of the docker images on dockerhub to point to dockerfiles from github
- tested the publish to my own nuget/choco server
- tested publish to rubygem (https://rubygems.org/gems/gitversion there is already published)

Right now I'm finilizind the TFS publishing so could you wait with the merge when I finish with the changes @gep13 ?

**Comment by gep13** [comment]
>@arturcic said...
Right now I'm finilizing the TFS publishing so could you wait with the merge till I finish with the changes @gep13? Hopefully today I'll have everything done.

Sounds good to me!

**Comment by asbjornu** [comment]
Big :+1: from me!

**Comment by arturcic** [comment]
I have tested the changes, the only thing that is not working - Publish-TFS task - is because of the personal access token expired when trying to publish to visual studion marketplace. There is a flag on appveyor that allows disabling the publishing for now for TFS till we get the new personal access token in place, so that we can merge the PR and see how the new scripts work. I have tested the Publish-TFS with a different name of the extension for testing purposes and it was working.

So basically the PR is ready for merge.

**Comment by dazinator** [comment]
@arturcic does this change sit ok with you: https://github.com/GitTools/GitVersion/commit/2cf0fb1

**Comment by arturcic** [comment]
Yeah I guess it is ok for the time we merge that branch. We'll just add the version of the sdk directly in the bootstrap script.

**Comment by kll** [comment]
I went and actually read the [docs](https://docs.microsoft.com/en-us/dotnet/core/tools/global-json) now and I think we can just specify the major.minor version to pin us to the 2.1 series which is LTS and probably a reasonable thing to do without getting tied to a specific release that quickly gets out of date. If it works like I think is that an agreeable middle ground?

**Comment by dazinator** [comment]
@kll that sounds good.

**Comment by kll** [comment]
Never mind. It definitely did not work like I thought it did. I also didn't realize this PR was reading in the contents of the file as part of the new build process so I've just added it back with a matching version number. Even with a specific version number it looks like it rolls forward on the patch version automatically which prevents needing to go find an old exact version number to make it build. Hopefully it's kept up to date because I rarely have old versions of the SDK installed. :)

**Comment by jmezach** [comment]
Is this still waiting for something in particular, or is this ready to merge?

**Comment by asbjornu** [comment]
@jmezach: This is such a giant leap in the right direction that regardless of what's missing, I'm going ahead and merging this. Thanks for the poke!

File: run.ps1 [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+.PARAMETER ScriptArgs
+Remaining arguments are added here.
+
+.LINK
+https://cakebuild.net
+#>
+
+[CmdletBinding()]
+Param(
+    [string]$Script = "build.cake",
```
Comment (kll) [inline run.ps1]: Shouldn't this run the new run.cake script by default? The old build.cake also seems to delete files needed by the new build script and leads to some confusion.

File: run.ps1 [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    [Parameter(Position=0,Mandatory=$false,ValueFromRemainingArguments=$true)]
+    [string[]]$ScriptArgs
+)
+
+Write-Host "Preparing to run build script..."
+
+$CakeVersion = "0.30.0"
+
+$DotNetChannel = "Current";
+$DotNetInstaller = if ($IsWindows) { "dotnet-install.ps1" } else { "dotnet-install.sh" }
```
Comment (kll) [inline run.ps1]: Where does $IsWindows come from? When I run this on Windows from PowerShell it evaluates to false which ultimately results in the following output which seems to be ignored because the script continues on successfully after.
`/bin/bash: C:UsersworkcodeGitVersion.dotnet/dotnet-install.sh: No such file or directory`

**Comment by arturcic** [comment]
I think I'll remove or probably rename back to `build.cake` and `build.ps1` after the #1445 is closed
